---
title: "RNA-seq Analysis"
subtitle: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  bookdown::pdf_document2:
    number_sections: true
    toc: true
    toc_depth: 1
    df_print: kable
    dev: cairo_pdf
    latex_engine: xelatex
geometry: margin = 0.5in
fontsize: 11pt
mainfont: Calibri
---

```{r setup, include = FALSE}
# load libraries
suppressPackageStartupMessages({
  devtools::load_all()
  library(tidyverse)
  library(wmo)
  library(targets)
})

# resolve conflicts
conflicted::conflict_prefer("filter", "dplyr")

# set global chunk options
knitr::opts_chunk$set(
  echo = FALSE, 
  message = FALSE, 
  warning = FALSE, 
  fig.align = "center", 
  out.width = "49%"
) 

options(knitr.table.format = function() {
  if (knitr::is_latex_output()) 
    "latex" else "html"
})

theme_set(theme_wmo(base_family = "Calibri"))
```

```{r}
withr::with_dir(here::here(), {
  pca <- tar_read(rnaseq_pca)
  deg <- tar_read(rnaseq_different_differences)
  results <- tar_read(rnaseq_different_differences)
  gsea_res <- tar_read(rnaseq_gsea)
  gsea_p <- tar_read(rnaseq_gsea_plot)
  tfea_p <- tar_read(rnaseq_tfea_plot)
  dds <- tar_read(dds_symbols)
})
```

\newpage

# Overview

We observed that proliferating primary cells exposed to hypoxia do not increase glucose uptake and lactate efflux despite up-regulation of glucose transporters and glycolytic genes. When these cells are treated with the prolyl hydroxylase inhibitor molidustat in normoxia, the expected increases in glycolytic flux are observed. Interestingly, when molidustat-treated cells are cultured in hypoxia, hypoxia blocks molidustat-mediated increases in glycolysis. In an effort to identify the mechanism mediating this effect, we performed RNA-seq on lung fibroblasts treated for three days with 0.5% oxygen or molidustat (10 Î¼M) with 21% and DMSO (0.1%) controls. RNA was collected and purified and sent for RNA-seq analysis.

# Analysis

We will use the `DESeq2` package to identify differentially expresssed genes. We will use the model `~ experiment + oxygen * treatment`. 

## PCA

```{r}
pca
```

## Differential Expression

In this analysis, will identify those genes where the effects of hypoxia and molidustat are different, which will hopefully point to mechanisms by which hypoxia may regulate cell metabolism differently than PHD2 inhibition.

```{r}
deg %>% 
  dplyr::select(row, symbol, description, lg2fc = log2FoldChange, padj) %>% 
  head(20) %>% 
  my_kable()
```

## Gene Set Enrichment Analysis

```{r}
gsea_p
```

```{r, eval = FALSE}
pathways <-
  fgsea::gmtPathways("~/Dropbox (Partners HealthCare)/msigdb_v7.2/msigdb_v7.2_GMTs/msigdb.v7.2.entrez.gmt")

df <- 
  gsea_res %>% 
  dplyr::filter(source == "HALLMARK") %>% 
  dplyr::filter(padj < 0.05) %>% 
  tidyr::unnest(leadingEdge) %>% 
  dplyr::inner_join(results, by = c("leadingEdge" = "row")) %>% 
  dplyr::group_by(pathway) %>% 
  dplyr::slice_max(order_by = stat, n = 1) %>% 
  dplyr::arrange(desc(NES))

gois <- df$symbol %>% rlang::set_names(df$pathway)

rnks <-
  results %>%
  dplyr::select(row, stat) %>%
  dplyr::arrange(stat) %>%
  tibble::deframe()

gsea_res <-
    fgsea::calcGseaStat(
      rnks,
      selectedStats = path,
      returnAllExtremes = TRUE
    )

fgsea::plotEnrichment(
  pathway = pathways[["HALLMARK_OXIDATIVE_PHOSPHORYLATION"]], 
  stats = rnks
)
```


```{r}
plot_rnaseq_goi(dds, "HIRA")
```


## Transcription Factor Analysis

```{r}
tfea_p
```
