---
title: "Hypoxia uncouples HIF gene transcription and metabolic flux in proliferating primary cells"

author:
- Courtney A. Copeland: 
    institute: 1
- Benjamin A. Olenchock: 
    institute: 1, 2
- David R. Ziehr:
    institute: 1, 4
- Sarah McGarrity: 
    institute: 1, 5
- Jamey D. Young: 
    institute: 3
- Joseph Loscalzo: 
    institute: 1
- William M. Oldham: 
    institute: 1
    email: woldham@bwh.harvard.edu
    correspondence: true
institute:
- 1: Department of Medicine, Brigham and Women's Hospital and Harvard Medical School, Boston, MA
- 2: Regeneron Pharmaceuticals, Tarrytown, NY
- 3: Departments of Chemical & Biomolecular Engineering and Molecular Physiology & Biophysics, Vanderbilt University, Nashville, TN
- 4: Department of Medicine, Massachusetts General Hospital and Harvard Medical School, Boston, MA
- 5: Center for Systems Biology, School of Health Sciences, University of Iceland, Reykjavik, Iceland

bibliography_main: library.bib
bibliography_software: packages.bib
csl: cell-metabolism.csl
link-citations: yes

output: 
  # bookdown::pdf_document2:
  #   # template: template.tex
  #   keep_tex: false
  #   df_print: kable
  #   toc: false
  #   number_sections: false
  #   latex_engine: xelatex
  #   pandoc_args:
  #     - '--lua-filter=scholarly-metadata.lua'
  #     - '--lua-filter=author-info-blocks.lua'
  #     - '--lua-filter=pagebreak.lua'
  #     - '--lua-filter=multiple-bibliographies.lua'
      
  bookdown::word_document2:
    reference_docx: template.docx
    df_print: kable
    fig_caption: true
    number_sections: false
    pandoc_args:
      - '--lua-filter=scholarly-metadata.lua'
      - '--lua-filter=author-info-blocks.lua'
      - '--lua-filter=pagebreak.lua'
      - '--lua-filter=multiple-bibliographies.lua'

compact-title: true
geometry: margin = 0.75in
fontsize: 11pt
mainfont: Calibri
graphics: yes
header-includes:
- \usepackage{lineno}
- \linenumbers
- \setlength\linenumbersep{1cm}
- \usepackage{setspace}
- \doublespacing
- \makeatletter\renewcommand*{\fps@figure}{htbp}\makeatother
- \setlength{\parskip}{0.75\baselineskip plus 2pt minus 2pt}
- \usepackage{float}
---

```{r setup, include = FALSE}
# library(Copeland.2021.hypoxia.flux)
devtools::load_all()
library(magrittr)

knitr::opts_chunk$set(
  echo = FALSE, 
  warning = FALSE, 
  message = FALSE, 
  out.width = "50%", 
  fig.align = "center", 
  fig.pos = "H", 
  out.extra = ""
)

knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

knitr::opts_knit$set(kable.force.latex = FALSE)

options(tinytex.verbose = TRUE)

# kable function
my_kable <- function(data, ...) {
  kableExtra::kable(data, booktabs = TRUE, linesep = "", ...) %>% 
    kableExtra::kable_styling(latex_options = c("HOLD_position"), font_size = 11)
}

withr::with_dir(here::here(), {
  biomass <- targets::tar_read(biomass)
  resources <- targets::tar_read(resources_table)
})

biomass <- tibble::deframe(biomass)
```

\newpage 

# Summary 
(150 words)

# Keywords

hypoxia, metabolic flux analysis, hypoxia-inducible factor, prolyl hydroxylase, metabolism

\newpage

# Introduction

Metazoan cells depend on aerobic respiration to meet cellular energy demands. With an inadequate oxygen supply, or hypoxia, cells must reduce energy consumption and shift energy production away from oxidative phosphorylation. Cells accomplish this goal through stabilization of the hypoxia-inducible transcription factor 1α (HIF-1α), which activates the transcription of glucose transporters, glycolytic enzymes, lactate dehydrogenase, and pyruvate dehydrogenase kinase, while decreasing the expression of proteins in the tricarboxylic acid (TCA) cycle and electron transport chain [@RN1980, @RN1198]. Overall, these changes in gene transcription should increase glycolytic capacity and divert glucose-derived pyruvate from oxidative phosphorylation toward lactate fermentation to maintain energy production and minimize the formation of reactive oxygen species [@RN2525]. 

While this "glycolytic shift" of primary carbon metabolism is well-described, the effects of hypoxia on other metabolic pathways are an area of active investigation [@RN953, @RN1603, @RN1198]. Since hypoxia is a prominent feature of cancer biology as tumor growth outstrips blood supply, most detailed metabolic studies of cell metabolism in hypoxia have used tumor cell models [@RN2395, @RN1447, @RN1438, @RN634, @RN1200, @RN975]. For example, stable isotope tracing and metabolic flux analyses identified a critical role for reductive carboxylation of glutamine-derived α-ketoglutarate for lipid biosynthesis in tumor growth [@RN1447; @RN628; @RN1966; @RN2395], and metabolomic studies identified aspartate as a limiting metabolite for cancer cell proliferation under hypoxia [@RN634]. By contrast, comparatively little is known about metabolic adaptations of primary cells to hypoxia and the importance of the metabolic pathways described above remain to be elucidated in these cells. This information would provide important context for understanding the extent to which cancer cell metabolism responds differently to hypoxic stress. Given the metabolic adaptations required for rapid proliferation in cancer cells, we hypothesized that hypoxia would elicit different metabolic responses in primary cells than has been observed previously in studies of cancer cell metabolism. 

To test this hypothesis, we developed models of bioenergetic carbon flux in human primary cells cultured under 21% or 0.5% oxygen conditions. We found that hypoxia fails to increase glycolysis in primary cells despite robust up-regulation of the HIF-1α transcriptional program. In normoxia, HIF-1α activation by the prolyl hydroxylase inhibitor molidustat (BAY-85-3934, "BAY") [@RN580] did increase glycolysis and lactate efflux; however, hypoxia abrogated this response. These findings suggest the existence of important hypoxia-dependent regulatory mechanisms that override the metabolic consequences of HIF-1α-dependent up-regulation of glycolytic gene expression in human primary cells. 

# Results

The goal of this study was to identify the metabolic changes associated with hypoxia in proliferating primary cells. Cells were seeded and placed into hypoxia for 24 h prior to sample collection to provide adequate time for activation of the hypoxia-dependent transcriptional program. From this starting point, we identified the optimal cell seeding density and time course to capture exponential cell growth (Figure \@ref(fig:f1)A), thought to be an indicator of metabolic steady state. Lung fibroblasts (LFs) cultured in 0.5% oxygen grew more slowly (Figure \@ref(fig:f1)B), but slower growth was not associated with decreased cell viability (Figure \@ref(fig:f1)C). These cells demonstrated robust stabilization of HIF-1α protein associated with up-regulation of downstream targets, such as glucose transporter 1, (GLUT1), and lactate dehydrogenase A (LDHA) (Figure \@ref(fig:f1)D-H). These changes persisted for the duration of the experimental time course.

## Extracellular flux analysis reveals little impact of hypoxia on glycolysis

Having established a model system, we next determined the extracellular fluxes of glucose (GLC), lactate (LAC), pyruvate (PYR), and amino acids (Figure \@ref(fig:f1)I-J, Supplementary Figures 1-3). Flux calculations incorporated the changes in cell number, extracellular metabolite concentrations, and medium evaporation over time [@RN1523] (Supplementary Figure 1). Surprisingly, while glucose uptake was modestly increased in hypoxia, lactate efflux was decreased (Figure \@ref(fig:f1)I) despite activation of the HIF-1 transcriptional program. Similar findings were observed when the ambient oxygen level was decreased further to 0.2% (Supplementary Figure 2) and in pulmonary artery smooth muscle cells (PASMCs; Supplementary Figure 3). In addition to lactate, the absolute extracellular fluxes of pyruvate and amino acids were generally decreased in hypoxia, including a marked decrease in glutamine uptake in LFs. Notably, hypoxia was previously shown to increase glutamine uptake in studies of cancer cell metabolism [@RN1447; @RN2395; @RN628]. Glutamine uptake did increase in PASMCs, as did the uptake of branch chain amino acids and arginine (Supplementary Figure 3), suggesting differential responses of these mesenchymal cells to hypoxia. 

Given that hypoxia did not increase glucose and lactate fluxes as expected from the associated increases in glycolytic gene expression, we next assessed the capacity of HIF-1α to augment glycolysis in LFs. Cells were treated with BAY to stabilize HIF-1α under 21% oxygen conditions (Figure \@ref(fig:f2). Similar to hypoxia, BAY decreased cell growth rate (Figure \@ref(fig:f2)A-B) and activated the HIF-1 transcriptional program (Figure \@ref(fig:f2)C-G). Compared to hypoxia, BAY treatment resulted in a similar activation of HIF-1 target gene transcription and protein expression. In normoxia, this transcriptional program was associated with substantially increased glucose uptake and lactate efflux (Figure \@ref(fig:f2)H). Comparatively modest effects of BAY on amino acid fluxes were observed as compared to 0.5% oxygen culture conditions (Figure \@ref(fig:f2)J).

## Stable isotope incorporation suggests preservation of the intracellular metabolic program in hypoxia

To validate our findings from extracellular flux measurements, we next treated LFs with stable carbon isotopes of glucose and glutamine to measure the rate of label incorporation and trace its incorporation into key carbon utilization pathways (Figure \@ref(fig:f3) and Supplementary Figures 4-5). Consistent with our extracellular flux measurements, we found that the rate of label incorporation from [U-^13^C~6~] glucose into pyruvate was similar across 21%, 0.5%, and DMSO-treated conditions, but markedly increased with BAY treatment (\@ref(fig:f3)A-B).

Overall, hypoxia-treated cells generally had decreased label incorporation into downstream metabolites (*i.e.*, the unlabeled, or M0, fraction was greater). This finding is consistent with the extracellular flux measurements suggesting slower substrate utilization by hypoxic cells. BAY treatment recapitulated the labeling patterns observed with hypoxia, suggesting similar effects on intracellular metabolite distribution between these two conditions. Beyond this observation, the labeling patterns in hypoxia- and BAY-treated cells were similar to their respective controls, arguing against marked metabolic reprogramming in response to prolyl hydroxylase inhibition by either hypoxia or BAY. These findings were similar in PASMCs (Supplementary Figure 5) where stable isotope incorporation was attenuated in hypoxia and the overall labeling pattern was quite similar in both conditions. Compared to previous studies of metabolic flux in cancer cells [@RN2395; @RN1447], no marked increase in M5-labeled citrate from [U-^13^C~5~] glutamine was observed with hypoxia treatment, indicating no increase in reductive carboxylation for lipid synthesis. The overall fraction of M5-citrate in these cells was low (< 6%). 

## Metabolic flux in hypoxia is closely coupled to cell growth rate

To clarify changes in intracellular metabolite fluxes, we next generated metabolic flux models incorporating the extracellular flux measurements and stable isotope tracing data described above. Preliminary labeling time courses indicated that, even after 72 h of labeling, intracellular metabolites did not reach isotopic steady state (Figure \@ref(fig:f3A), Supplementary Figure 6). Thus, we performed isotopically non-stationary metabolic flux analysis as implemented by INCA [@RN1522; @RN2501; @RN964] (Figure \@ref(fig:f4), Supplementary Figure 6, and Supplementary Tables 1-3). Overall, proliferating LFs demonstrated high rates of glucose uptake and glycolysis. Approximately 10% of cytoplasmic pyruvate enters the TCA cycle with the balance converted to lactate (Supplementary Figure 6C). In hypoxia, significant reductions in glycolysis, the TCA cycle, and amino acid metabolism were observed (Figure \@ref(fig:f4)A) with a significant increase in pentose phosphate pathway flux. 

Given the global decrease in bioenergetic metabolic flux in hypoxia, we hypothesized that these differences may be a consequence of a decreased growth rate. After normalizing metabolite fluxes in normoxia and hypoxia to the associated growth rate, modest increases (~10%) in glycolytic flux were observed (Supplementary Figure 4D). This finding suggests that, while glycolysis increases relative to growth rate in hypoxic cells, the regulators of cell proliferation rate dominate over the consequences of the HIF-1 transcriptional program. Indeed, even after adjusting for cell growth rate, the relative increase in glycolytic flux is modest compared to the marked up-regulation of glycolytic genes.  

By comparison, HIF-1 activation by BAY in 21% oxygen increased glycolysis and lactate fermentation by ~50% (Figure \@ref(fig:f4)B). BAY treatment decreased cell proliferation rate (Figure \@ref(fig:f2)B), indicating that, unlike hypoxia, BAY treatment uncouples cell proliferation and metabolic flux. Similar to hypoxia, decreases in serine and glutamine incorporation were observed. Metabolite fluxes in DMSO-treated cells were similar to 21% oxygen controls. 

## Hypoxia and BAY treatment increase lactate oxidation

Although the metabolite exchange fluxes for bidirectional reactions tend to be poorly resolved, two observations are worth highlighting (Supplementary Tables 1 and 2). First, consistent with the stable isotope tracing results, the rate of reductive carboxylation through reversible flux by isocitrate dehydrogenase is low (~4 fmol/cell/h), is unchanged by hypoxia, and is modestly increased by BAY treatment. 

Second, hypoxia and BAY treatment are associated with a marked increase in the lactate transport exchange flux in LFs from ~ 0 to 1,520 and 711 fmol/cell/h in 0.5% oxygen and BAY treatment conditions, respectively. Since the net lactate transport flux is secretion, this observation suggests increased lactate uptake with hypoxia or BAY treatment. This may be consistent with the HIF-driven increased expression of the reversible lactate transporter MCT4 [@RN366]. To investigate this hypothesis, LFs and PASMCs were treated with [U-^13^C~3~] lactate (2 mM) and ^13^C incorporation into intracellular metabolites was analyzed by LC-MS (Figure \@ref(fig:f5), Supplementary Figure 5). Here, we observed increased labeling of TCA metabolites citrate (CIT), 2-oxoglutrate (2OG), malate (MAL), and aspartate (ASP) following hypoxia or BAY treatment (Figure \@ref(fig:f5) and Supplementary Figure 4). Interestingly, a similar increase in labeling was not observed in PASMCs (Supplementary Figure 5). 

Notably, lactate labeled ~50% of citrate and ~20% of downstream TCA cycle metabolites in both LFs and PASMCs, indicating that lactate may be an important respiratory fuel source in these cells even though lactate efflux is high. Although lactate has been used less commonly than glucose and glutamine in stable isotope tracing studies, Faubert and colleagues [-@RN2600] demonstrated lactate incorporation in human lung adenocarcinoma *in vivo*. In this study, lactate incorporation corresponded to regions of high glucose uptake as determined by `[`^18^F]-fluorodeoxyglucose positron emission tomography, suggesting that lactate consumption can occur in areas of high glucose utilization. Subsequently, several investigators have demonstrated the importance of lactate as a metabolic fuel *in vivo* [@RN2756; @RN2718]. In addition to downstream metabolites, we also observed hypoxia- and BAY-dependent increases in lactate incorporation in fructose bisphosphate (FBP) and 3-phosphoglycerate (3PG). This is consistent with prior reports describing hypoxia-mediated increases in glycogen synthesis [@RN1678; @RN545; @RN1668]. Together, these data suggest that lactate makes a modest (~5% carbon) contribution to this process.

## Hypoxia abrogates the increase in glycolysis following pharmacologic prolyl hydroxylase inhibition

To reconcile the differential effects of prolyl hydroxylase inhibition by hypoxia and BAY, we next addressed whether hypoxia could suppress the effects of BAY on glucose and lactate fluxes (Figure \@ref(fig:f6)A-C). LFs cultured in standard growth medium were treated with BAY and placed in either 21% or 0.5% oxygen. Similar to previous experiments, BAY treatment decreased cell growth rate, increased glucose uptake, and increased lactate efflux in 21% oxygen. Interestingly, when combined with 0.5% oxygen, BAY treatment was unable to enhance lactate efflux. These data suggest that hypoxia antagonizes the effects of HIF-1 activation on glycolytic flux in these primary cells. 

To investigate these metabolic differences further, we performed metabolomic profiling of LFs treated for 72 h with hypoxia or BAY separately or in combination (Figure \@ref(fig:f6)D-F). Principle component analysis revealed greater class similarity among both treatment groups cultured in 0.5% oxygen (Figure \@ref(fig:f6)D). These hypoxia-treated cells were well-segregated from BAY-treated cells, again highlighting differential effects of hypoxic and pharmacologic PHD inhibition. Linear modeling identified those metabolites that were differentially affected by hypoxia and BAY treatment (*i.e.*, those metabolites with a significant interaction term in the linear model described by treatment × oxygen) (Figure \@ref(fig:f6)E). Several patterns emerged from this analysis (Figure \@ref(fig:f6)F). Interestingly, the hypoxia-regulated metabolite, 2-hydroxyglutarate, increased following BAY treatment, but was unaffected by hypoxia treatment in these proliferating LFs [@RN1603]. The tricarboxylic acid cycle metabolite aconitate was decreased in hypoxia, but increased by BAY treatment while the opposite effect was observed for hydroxyproline and taurine. Glycolytic intermediate glyceraldehyde 3-phosophate (GAP) was decreased by both hypoxia and BAY, while hypoxia reversed the BAY-mediated decrease in γ-aminobutyric acid (GABA). A metabolite set enrichment analysis revealed the tricarboxlyic acid cycle to be the most

One of the key differences associated with 



# Discussion

# Acknowledgements

# Author Contributions

# Declaration of Interests

\newpage 

# Figure Legends

(ref:f1) **Extracellular fluxes of lung fibroblasts in hypoxia.** (**A**) Growth curves of lung fibroblasts (LFs) cultured in 21% and 0.5% oxygen (n = 8). (**B**) Growth rates were determined by linear fitting of log-transformed growth curves. (**C**) Cell viability, assessed by acridine orange plus propidium iodide staining, did not differ between 21% and 0.5% oxygen culture conditions (n = 3 technical replicates). (**D**) Representative immunoblot of LF protein lysates collected at the indicated times. (**E**, **F**) Relative change in HIF-1α (E) and LDHA (F) protein levels compared to 21% oxygen time 0 (n = 4). (**G**, **H**) Relative changes in GLUT1 (G) and LDHA (H) mRNA levels compared to 21% oxygen time 0 (n = 4). (**I**, **J**) Extracellular fluxes of the indicated metabolites (n = 8). Data are mean ± SEM. Comparisons were performed using Student's paired *t*-test and p < 0.05 indicated by *. Flux probability values were not corrected for multiple comparisons. Abbreviations as noted in the text.

(ref:f2) **Prolyl hydroxylase inhibition of lung fibroblasts in normoxia.** (**A**) Growth curves of lung fibroblasts (LFs) cultured in 21% oxygen and treated with molidustat (BAY, 10 μM) or vehicle (DMSO, 0.1%) (n = 8). (**B**) Growth rates were determined by linear fitting of log-transformed growth curves. (**C**) Representative immunoblot of LF protein lysates collected at the indicated times. (**D**, **E**) Relative change in HIF-1α (D) and LDHA (E) protein levels compared to DMSO time 0. (**F**, **G**) Relative changes in GLUT1 (F) and LDHA (G) mRNA levels compared to DMSO time 0. (**H**, **I**) Extracellular fluxes of the indicated metabolites (n = 8). Data are mean ± SEM. Comparisons were performed using Student's paired *t*-test and p < 0.05 indicated by *. Flux probability values were not corrected for multiple comparisons. Abbreviations as noted in the text.

(ref:f3) **Stable isotope tracing of lung fibroblasts.** (**A**) Fraction of pyruvate labeling following treatment of lung fibroblasts (LFs) with [U-^13^C~6~] glucose. (**B**) Rate values determined from asymptotic regression fit of the data from (A). (**C**) Isotopic labeling of key intracellular metabolites after 72 h of treatment with the indicated tracers. Data are the mean ± SEM of 4 biological replicates. FBP, fructose bisphosphate; 3PG, 3-phosphoglycerate; AKG, α-ketoglutarate.

(ref:f4) **Metabolic flux maps of lung fibroblasts.** (**A**) Ratio of metabolic fluxes in 0.5% oxygen compared to 21% oxygen. (**B**) Ratio of metabolic fluxes in cells treated with molidustat (BAY) compared to DMSO vehicle control. Fluxes with non-overlapping confidence intervals are highlighted to indicate significant changes.

(ref:f5) **PHD inhibition increases lactate uptake and oxidation.** Mass isotopomer distributions of key metabolites following labeling with [U-^13^C~3~] lactate (2 mM) for 72 h indicates increased lactate uptake and oxidation in hypoxia or with molidustat (BAY) treatment.

(ref:f6) **Metabolomic analysis of molidustat treamtent in normoxia and hypoxia.** (**A**-**C**) Hypoxia inhibits the effects of HIF-1α stabilization on glycolysis. Lung fibroblasts (LFs) were cultured in standard growth medium and treated with molidustat (BAY, 10 μM) or vehicle (DMSO, 0.1%) in 21% or 0.5% oxygen conditions (n = 4). (A) Growth rates were determined by linear fitting of log-transformed growth curves. (B, C) Extracellular fluxes of glucose (B) and lactate (C). (**D**) Principle components analysis of intracellular metabolites following 72 h of treatment described above suggests a dominant effect of hypoxia over PHD inhibition on the metabolome (n = 4). (**E**) Volcano plot of the differential effects of hypoxia and molidustat (BAY) treatment on intracellular metabolites. (**F**) Representative metabolites from (E) reveal different patterns of metabolic effects. (**G**-**I**) Intracellular NAD^+^, NADH, and their ratio were determined by enzymatic assay (n = 5). Black * indicate a significant effect of hypoxia, colored * indicate a significant effect of treatment within a given oxygen tension as indicated by the color. Comparisons for fluxes (A-C) and NAD(H) measurements (G-I) were performed using a mixed-effects linear model with replicate as a random effect. Adjusted p-values for the indicated comparisons were determined using Tukey's *post hoc* test.

(ref:f7) **Transcriptomic analysis of molidustat treamtent in normoxia and hypoxia.** (**A**) Principle components analysis of transcriptional changes following 72 h of treatment with 0.5% oxygen or molidustat (BAY), separately or together (n = 4). (**B**) Volcano plot of the differential effects of hypoxia and molidustat (BAY) treatment on intracellular metabolites. (**C**) Representative transcripts from (B). (**D**) Gene set enrichment analysis of trasncripts from (B). (**E**) Transcription factor enrichment analysis suggests mechanisms for differential regulation of gene expression following hypoxia or BAY treatment. (**F**) Relative change in c-MYC protein levels after 72 h of treatment (n = 4) and a representative immunoblot. Comparisons were performed using a mixed-effects linear model with replicate as a random effect. Adjusted p-values for the indicated comparisons were determined using Tukey's *post hoc* test. Black * indicate a significant effect of treatment, colored * indicate a significant effect of oxygen within a given treatment as indicated by the color.

Figure 1: (ref:f1)

Figure 2: (ref:f2)

Figure 3: (ref:f3)

Figure 4: (ref:f4)

Figure 5: (ref:f5)

Figure 6: (ref:f6)

Figure 7: (ref:f7)

\newpage

# Tables

\newpage

# Figures

```{r f1, fig.cap = "(ref:f1)"}
knitr::include_graphics(
  system.file(
    "manuscript/figures/m1.pdf",
    package = "Copeland.2021.hypoxia.flux"
  )
)
```

\newpage

```{r f2, fig.cap = "(ref:f2)"}
knitr::include_graphics(
  system.file(
    "manuscript/figures/m2.pdf",
    package = "Copeland.2021.hypoxia.flux"
  )
)
```

\newpage

```{r f3, fig.cap = "(ref:f3)"}
knitr::include_graphics(
  system.file(
    "manuscript/figures/m3.pdf",
    package = "Copeland.2021.hypoxia.flux"
  )
)
```

\newpage

```{r f4, fig.cap = "(ref:f4)"}
knitr::include_graphics(
  system.file(
    "manuscript/figures/m4.pdf",
    package = "Copeland.2021.hypoxia.flux"
  )
)
```

\newpage

```{r f5, fig.cap = "(ref:f5)"}
knitr::include_graphics(
  system.file(
    "manuscript/figures/m5.pdf",
    package = "Copeland.2021.hypoxia.flux"
  )
)
```

\newpage

```{r f6, fig.cap = "(ref:f6)"}
knitr::include_graphics(
  system.file(
    "manuscript/figures/m6.pdf",
    package = "Copeland.2021.hypoxia.flux"
  )
)
```

\newpage

```{r f7, fig.cap = "(ref:f7)"}
knitr::include_graphics(
  system.file(
    "manuscript/figures/m7.pdf",
    package = "Copeland.2021.hypoxia.flux"
  )
)
```

\newpage

# STAR Methods

## Resource Availability

### Lead Contact

Further information and requests for resources and reagents should be directed to and will be fulfilled by the lead contact, William Oldham (woldham@bwh.harvard.edu). 

### Materials Availability

This study did not generate new unique reagents.

### Data and code availability

The original data and analysis code are available as a reproducible research compendium formatted as an R package available at OldhamLab/Copeland.2021.hypoxia.flux (https://github.com/oldhamlab/Copeland.2021.hypoxia.flux). RNA-seq data has been deposited in the NIH Short Read Archive (PRJNA721596). 

## Experimental Model Details

### Lung fibrolbasts

Primary normal human lung fibroblasts (LFs) were purchased from Lonza (CC-2512) and cultured in FGM-2 (Lonza CC-3132) in a standard tissue culture incubator in 5% CO~2~ at 37 °C. Cells from two donors were used in these studies: #33652 (56, male) and #29132 (19, female). Cell authentication was performed by the vendor.

### Pulmonary artery smooth muscle cells

Primary human pulmonary artery smooth muscle cells were purchased from Lonza (CC-2581) and cultured in SmGM-2 (Lonza CC-3182) in a standard tissue culture incubator in 5% CO~2~ at 37 °C. Cells from multiple donors were used in these studies: #30020 (64, male), #27662 (35, male), #26698 (51, male), and #19828 (51, male). Cell authentication was performed by the vendor.

## Method Details

### Metabolic flux protocol

For extracellular flux measurements, cells were seeded in either standard growth medium or MCDB131 medium without glucose, glutamine, or phenol red (genDEPOT) supplemented with 2% dialyzed fetal bovine serum (Mediatech) and naturally labeled glucose and glutamine ("light" labeling medium). For LFs, glucose was supplemented at 8 mM and glutamine was supplemented at 1 mM. For PASMCs, glucose was supplemented at 5.55 mM and glutamine was supplemented at 10 mM. These concentrations match the concentrations of these substrates determined in standard growth medium. Preliminary experiments were performed to identify the optimal cell seeding density, exponential growth phase, and labeling duration consistent with metabolic and isotopic steady state. On Day -1, 25,000 cells were seeded in a 35 mm dish in "light" labeling medium. Hypoxia-treated cells were transferred to a tissue culture glovebox set to 0.5% oxygen and 5% CO~2~ (Coy Lab Products). Medium was supplemented with DMSO 0.1% or BAY (10 μM) for DMSO and BAY treatment conditions. On Day 0, cells were washed with PBS and the medium was changed to either "light" medium for flux measurements or "heavy" medium for tracer experiments. For LFs, samples were collected on Day 0 and every 24 h for 72 h. For PASMCs, samples were collected on Day 0 and every 12 h for 48 h. Medium and cell lysates were collected at each time point for intra- and extracellular metabolite measurements and total DNA quantification. Dishes without cells were weighed daily to correct for evaporative medium losses and to empirically determine degradation and accumulation rates of metabolites. Medium samples and cell lysates for DNA measurement were stored at -80 °C until analysis. Each individual experiment included triplicate wells for each treatment and time point, and each experiment was repeated 4-8 times. 

### Cell count

Direct cell counts of trypsizined cell suspensions in PBS were obtained following staining with propidium iodide and acridine orange using a LUNA-FL fluorescence cell counter (Logos Biosystems). Indirect cell counts for flux measurements were interpolated from total DNA quantified using the Quant-iT PicoGreen dsDNA Assay Kit (Thermo). Cells were washed once with two volumes of PBS, lysed with Tris-EDTA buffer containing 2% Triton X-100, and collected by scraping. Total DNA in 10 μL of lysate was determined by adding 100 μL of 1X PicoGreen dye in Tris-EDTA buffer and interpolating the fluorescence intensity with a standard curve generated using the λ DNA standard. Cell counts were interpolated from a standard curve of DNA obtained from known cell numbers seeded in basal medium (Supplementary Figure 1A). No difference in total cellular DNA was identified between normoxia and hypoxia cultures (Supplementary Figure 1B). 

### Immunoblots

Cells were washed with one volume of PBS and collected by scraping in PBS. Cell suspensions were centrifuged at 5,000 ×*g* for 5 min at 4 °C. Pellets were lysed in buffer containing Tris 10 mM, pH 7.4, NaCl 150 mM, EDTA 1 mM, EGTA 1 mM, Triton X-100 1% v/v, NP-40 0.5% v/v, and Halt Protease Inhibitor Cocktail (Thermo). Protein concentrations were determined by BCA Protein Assay (Thermo). Lysates were normalized for protein concentration and subjected to SDS-PAGE separation on stain-free tris-glycine gels (Bio-Rad), cross-linked and imaged with the Chemidoc system (Bio-Rad), transferred to PVDF membranes with the Trans-Blot Turbo transfer system (Bio-Rad), imaged, blocked in 5% blocking buffer (Bio-Rad), blotted in primary and secondary antibodies, and developed using WesternBright ECL (Advansta). Band signal intensity was normalized to total protein per lane as determined from the stain-free gel or membrane images. 

### RT-qPCR

Total RNA was isolated from cells with the RNeasy Mini Kit (Qiagen). cDNA was synthesized from 0.25-1.00 ng RNA with the High Capacity cDNA Reverse Transcription Kit (Applied Biosystems). RT-qPCR analysis was performed with an Applied Biosystems 7500 Fast Real Time PCR System with TaqMan Universal PCR Master Mix and pre-designed TaqMan gene expression assays (Life Technologies). Relative expression levels were calculated using the comparative cycle threshold method referenced to *ACTB*. 

###  Glucose assay

Medium samples were diluted 10-fold in PBS. Glucose concentration was determined using the Glucose Colorimetric Assay Kit (Cayman) according to the manufacturer's protocol. Standards were prepared in PBS.  

### Lactate assay

Medium samples were diluted 10-fold in PBS. Glucose concentration was determined using the ʟ-Lactate Assay Kit (Cayman). Medium samples did not require deproteinization, otherwise the samples were analyzed according to the manufacturer's protocol. Standards were prepared in PBS. 

### Pyruvate assay

Pyruvate was measured using either an enzymatic assay (most samples) or an HPLC-based assay (medium from 0.2% oxygen experiments). For the enzymatic assay, medium samples were diluted 20-fold in PBS. Pyruvate concentration was determined using the Pyruvate Assay Kit (Cayman). Medium samples did not require deproteinization, otherwise the samples were analyzed according to the manufacturer's protocol. Standards were prepared in PBS. For the HPLC assay, 2-oxovaleric acid was added to medium samples as an internal standard. Samples were subsequently deproteinized with 2 volumes of ice-cold acetone. Supernatants were evaporated to < 50% of the starting volume at 43 °C in a SpeedVac concentrator (Thermo Savant) and reconstituted to the starting volume with HPLC-grade water prior to derivatization. Samples were derivatized 1:1 by volume with *o*-phenylenediamine (25 mM in 2 M HCl) for 30 min at 80 °C. Derivatized pyruvate was separated with a Poroshell HPH C-18 column (2.1 × 100 mm, 2.7 μm) on an Infinity II high-performance liquid chromatography system with fluorescence detection of OPD-derivatized α-keto acids as described previously [@RN730].

### Amino acid assay

Medium amino acid concentrations were determined following the addition of norvaline and sarcosine internal standards and deproteinization with 2 volumes of ice-cold acetone. Supernatants were evaporated to < 50% of the starting volume at 43 °C in a SpeedVac concentrator (Thermo Savant) and reconstituted to the starting volume with HPLC-grade water prior to analysis. Amino acids in deproteinized medium were derivatized with *o*-phthalaldehyde (OPA) and 9-fluorenylmethylchloroformate (FMOC) immediately prior to separation with a Poroshell HPH-C18 column (4.6 × 100 mm, 2.7 μm) on an Infinity II high-performance liquid chromatography system with ultraviolet and fluorescence detection of OPA- and FMOC-derivatized amino acids, respectively, according to the manufacturer's protocol (Agilent) [@RN1294]. 

### Flux calculations

The growth rate ($\mu$) and flux (*v*) for each measured metabolite were defined as follows [@RN1523]:

\begin{align}
\frac{dX}{dt} &= \mu X (\#eq:1) \\[6 pt] 
\frac{dM}{dt} &= -kM+vX (\#eq:2) 
\end{align}

where *X* is the cell density, *k* is the first-order degradation or accumulation rate, and *M* is the mass of the metabolite. These equations are solved as follows:

\begin{align}
X &= X_0e^{\mu t} (\#eq:3)\\[6 pt]
Me^{k t} &= \frac{v X_0}{\mu + k} (e^{(\mu + k)t}-1) + M_0 (\#eq:4) 
\end{align}

Growth rate ($\mu$) and cell count at time 0 ($X_0$) were determined by robust linear modeling of the logarithm of cell count as a function of time ($t$). Metabolite mass was calculated from the measured metabolite concentrations and predicted well volume accounting for evaporative losses (Supplementary Figure 1C). First-order degradation and accumulation rates were obtained from robust linear modeling of metabolite mass *v*. time in unconditioned culture medium. Rates that significantly differed from 0 using Student's *t*-test were incorporated into the flux calculations. Final fluxes were obtained by robust linear modeling of $Me^{k t}$ *versus* $(e^{(\mu + k)t}-1)$ to determine the slope from which $v$ was calculated using equation \@ref(eq:4).

### Metabolomics

#### Metabolite extraction

Intracellular metabolites were obtained after washing cells with 2 volumes of ice-cold PBS and floating on liquid nitrogen. Plates were stored at -80 °C until extraction. Metabolites were extracted with 1 mL 80% MeOH pre-cooled to -80 °C containing 10 nmol [D~8~]-\textsc{dl}-valine as an internal standard (Cambridge Isotope Labs). Insoluble material was removed by centrifugation at 21,000 ×*g* for 15 min at 4 °C. The supernatant was evaporated to dryness at 42 °C using a SpeedVac concentrator (Thermo Savant). Samples were resuspended in 35 μL LC-MS-grade water prior to analysis.

#### Acquisition parameters

LC-MS analysis was performed on a Vanquish ultra-high-performance liquid chromatography system coupled to a Q Exactive orbitrap mass spectrometer by a HESI-II electrospray ionization probe (Thermo). External mass calibration was performed weekly. Metabolite samples (2.5 μL) were separated using a ZIC-pHILIC stationary phase (2.1 × 150 mm, 5 μm) (Merck). The autosampler temperature was 4 °C and the column compartment was maintained at 25 °C. Mobile phase A was 20 mM ammonium carbonate and 0.1% ammonium hydroxide. Mobile phase B was acetonitrile. The flow rate was 0.1 mL/min. Solvent was introduced to the mass spectrometer *via* electrospray ionization with the following source parameters: sheath gas 40, auxiliary gas 15, sweep gas 1, spray voltage +3.0 kV for positive mode and -3.1 kV for negative mode, capillary temperature 275 °C, S-lens RF level 40, and probe temperature 350 °C. Data were acquired and peaks integrated using TraceFinder 4.1 (Thermo). 

#### Stable isotope quantification

All metabolites except fructose 2,6-bisphosphate (FBP) and 3-phosphoglycerate (3PG) were measured using the following mobile phase gradient: 0 min, 80% B; 5 min, 80% B; 30 min, 20% B; 31 min, 80% B; 42 min, 80% B. The mass spectrometer was operated in selected ion monitoring mode with an m/*z* window width of 9.0 centered 1.003355-times half the number of carbon atoms in the target metabolite. The resolution was set at 70,000 and AGC target was 1 × 10^5^ ions. Peak areas were corrected for quadrupole bias as previously described [@RN1063]. Mass isotope distributions for FBP and 3PG were calculated from full scan chromatograms as described below. Raw mass isotopomer distributions were corrected for natural isotope abundance using a custom R package (mzrtools, https://github.com/oldhamlab/mzrtools) employing the method of Fernandez, *et al.* [@RN557]. 

#### Metabolomic profiling

For metabolomic profiling and quantification of isotopic enrichment for FBP and 3PG, the following mobile phase gradient was used: 0 min, 80% B; 20 min, 20% B; 20.5 min, 80% B; 28 min, 80% B; 42 min, 80% B. The mass spectrometer was operated in polarity switching full scan mode from 70-1000 m/*z*. Resolution was set to 70,000 and the AGC target was 1x10^6^ ions. Peak identifications were based on an in-house library of authentic metabolite standards previously analyzed utilizing this method. For metabolomics studies, pooled quality control (QC) samples were injected at the beginning, end, and between every four samples of the run. Raw peak areas for each metabolite were corrected for instrument drift using a cubic spline model of QC peak areas. Low quality features were removed on the basis of a relative standard deviation greater than 0.2 in the QC samples and a dispersion ratio greater than 0.4 [@RN3077]. Missing values were imputed using random forrest. Samples peak areas were normalized using probabilistic quotient normalization [@RN3094]. Differentially regulated metabolites were identified using limma [@limma]. 

### Biomass determination

The dry weight of LFs was determined to be `r round(biomass[["lf"]])` pg/cell. The dry weight of PASMCs was determined to be `r round(biomass[["pasmc"]])` pg/cell. These values were estimated by washing 3 × 10^6^ cells twice in PBS and thrice in ice-cold acetone prior to drying overnight in a SpeedVac. The composition of the dry cell mass was estimated from the literature [@RN1741; @RN2005], and stoichiometric coefficients were determined as described [@RN1522; @RN2517].

### Metabolic flux analysis

Metabolic flux analysis was performed using the elementary metabolite unit-based software package INCA [@RN2501]. Inputs to the model include the chemical reactions and atom transitions of central carbon metabolism, extracellular fluxes, the identity and composition of ^13^C-labeled tracers, and the MIDs of labeled intracellular metabolites. The metabolic network was adapted from previously published networks [@RN1522; @RN2266] and comprises `r nrow(dplyr::filter(model_reactions, !(pathway %in% c("mixing", "dilution"))))` reactions representing glycolysis, the pentose phosphate pathway, the tricarboxylic acid cycle, anaplerotic pathways, serine metabolism, and biomass synthesis. The network includes seven extracellular substrates (aspartate, cystine, glucose, glutamine, glycine, pyruvate, serine) and five metabolic products (alanine, biomass, glutamate, lactate, lipid). Models were fit using three ^13^C-labeled tracers, [1,2-^13^C~2~] glucose, [U-^13^C~6~] glucose, and [U-^13^C~5~] glutamine. The MIDs of twelve metabolites (2-oxoglutarate, 3-phosphoglycerate, alanine, aspartate, citrate, fructose bisphosphate, glutamate, glutamine, lactate, malate, pyruvate, serine) were used to constrain intracellular fluxes. The following assumptions were made:

1. Metabolism was at steady state.
1. Labeled CO~2~ produced during decarboxylation reactions left the system and did not re-incorporate during carboxylation reactions.
1. Protein turnover occurred at a negligible rate compared to glucose and glutamine consumption.
1. Acetyl-CoA, aspartate, fumarate, malate, oxaloacetate, pyruvate existed in cytosolic and mitochondrial pools. Aspartate and malate were allowed to exchange freely between the compartments.
1. The per cell biomass requirements of proliferating lung fibroblasts were similar to published estimated in other cells [@RN1741; @RN2005]. 
1. Dilution of alanine, aspartate, glutamate, glutamine, lactate, and pyruvate was allowed to occur through reversible exchange with unlabeled substrates in the medium as these metabolites were present in unconditioned medium or serum.
1. Succinate and fumarate are symmetric molecules that have interchangeable orientations when metabolized by TCA cycle enzymes. 

Flux estimation was repeated a minimum of 50 times from random initial values. Results were subjected to a χ^2^ statistical test to assess goodness-of-fit. Accurate 95% confidence intervals were computed for estimated parameters by evaluating the sensitivity of the sum-of-square residuals to parameter variations [@RN60; @RN1522]. 

### NAD(H) assay

Cellular NAD^+^ and NADH were measured using an enzymatic fluorimetric cycling assay based on the reduction of NAD^+^ to NADH by alcohol dehydrogenase (ADH) and subsequent electron transfer to generate the fluorescent molecule resorufin [@RN1603]. Briefly, cells were washed twice with one volume PBS. Pyridine nucleotides were extracted on ice with buffer containing 50% by volume PBS and 50% lysis buffer (100 mM sodium carbonate, 20 mM sodium bicarbonate, 10 mM nicotinamide, 0.05% by volume Triton-X-100, 1% by mass dodecyltrimethylammonium bromide) and collected by scraping. Extracts were divided equally and 0.5 volume of 0.4 N HCl was added to one sample. Both extracts were heated at 65 °C for 15 min to degrade selectively either the oxidized (buffer) or reduced (HCl) nucleotides. The reaction was cooled on ice and quenched by adding 0.5 M Tris-OH to the acid-treated samples or 0.2 N HCl plus 0.25 M Tris-OH to the buffer samples.  Samples were then diluted in reaction buffer (50 mM EDTA and 10 mM Tris, pH 7.06).  Cell debris was pelleted by centrifugation, and 50 μL was incubated for 2 h with 100 μL reaction buffer containing 0.6 M EtOH, 0.5 mM phenazine methosulfate, 0.05 mM resazurin, and 0.1 mg/mL ADH. Fluorescence intensities were measured with a Spectramax Gemini XPS (Moelcular Devices) with excitation 540 nm, emission 588 nm, and 550 nm excitation cut-off filter. Sample intensities were compared to a standard curve generated from known concentrations of NADH. The ratio of fluorescence in buffer-extracted to acid-extracted samples corresponds to the NADH/NAD^+^ ratio. Absolute NADH and NAD^+^ were normalized to cell count from cells estimated from total DNA quantification as described above.

### RNA-seq

RNA was collected from LFs treated for three days ± hypoxia ± BAY as described above. Four biological replicates were analyzed. Library construction and sequencing was performed by BGI Genomics using 100 bp paired end analysis and a read depth of 50M reads per sample. Sequences were deposited in the NIH SRA (PRJNA721596). Sequences were mapped to the human GRCh38 primary assembly and counts summarized using Rsubread [@Rsubread]. This data is available from the Oldham Lab GitHub repository (https://github.com/oldhamlab/rnaseq.lf.hypoxia.molidustat). Differentially expressed transcripts were identified using DESeq2 [@DESeq2]. Gene set enrichment and transcription factor enrichment was performed using the fgsea and TFEA.ChIP R packages, respectively [@fgsea; @TFEA.ChIP]. 

## Quantification and Statistical Analysis

The raw data and annotated analysis code necessary to reproduce this manuscript are contained in an R package research compendium available from the Oldham Lab GitHub repository (https://github.com/oldhamlab/Copeland.2021.hypoxia.flux). Data analysis, statistical comparisons, and visualization were performed in R [@R-base] using the packages listed below. Experiments included technical and biological replicates as noted above. The number of biological replicates (n) is indicated in the figure legends. Summary data show the mean ± SEM. Outliers were identified using twice the median absolute deviation. Two group comparisons (*e.g.*, 21% *v.* 0.5% oxygen) were performed using Student's *t*-test on log-transformed values with pairing among biological replicates. Multifactor comparisons (*e.g.*, BAY and oxygen) were performed using linear mixed-effects models with Tukey's *post hoc* testing. Metabolomics and RNA-seq data were analyzed as described above. Probability values less than 0.05 were considered significant.  

\newpage

## Key Resources Table  

<br>

```{r, results = "asis"}
flextable::flextable_to_rmd(resources)
```

\newpage

# Supplemental Items Titles

\newpage

# References

::: {#refs_main}
:::

# Software

::: {#refs_software}
:::
